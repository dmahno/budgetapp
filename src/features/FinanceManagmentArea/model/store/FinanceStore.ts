// FinanceStore.ts
import {makeAutoObservable} from 'mobx';
import {message} from 'antd';

import {getDatabase} from '../lib/getDatabase/getDatabase';

interface IOperation {
  id?: number;
  amount: number;
  description: string;
  type: 'add' | 'exclude';
  date: Date;
}

export class FinanceStore {
  operationsList: IOperation[] = [];

  constructor() {
    makeAutoObservable(this);
    this.loadOperations();
  }

  async loadOperations() {
    const database = await getDatabase();
    this.operationsList = await database.getAll('operations');
  }

  async addOperation(
    amount: number,
    description: string,
    type: 'add' | 'exclude',
  ) {
    if (type === 'exclude' && amount > this.totalSum) {
      message.error(
        `Недостаточно средств. Максимальная сумма для списания: ${this.totalSum} ₽`,
      );
      return;
    }

    const database = await getDatabase();
    const newOperation: IOperation = {
      amount,
      description,
      type,
      date: new Date(),
    };

    const operationId = (await database.add(
      'operations',
      newOperation,
    )) as number;
    this.operationsList.push({...newOperation, id: operationId});
  }

  async deleteOperation(operationId: number) {
    const database = await getDatabase();
    await database.delete('operations', operationId);
    this.operationsList = this.operationsList.filter(
      (operation) => operation.id !== operationId,
    );
  }

  get totalSum() {
    return this.operationsList.reduce((accumulatedSum, operation) => {
      return operation.type === 'add'
        ? accumulatedSum + operation.amount
        : accumulatedSum - operation.amount;
    }, 0);
  }
}

export const financeStore = new FinanceStore();
