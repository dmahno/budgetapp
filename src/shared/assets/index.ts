export type TImageKeys = '1.jpg' | '2.jpg' | '3.jpg' | '4.jpg' | '5.jpg';

import image1 from './images/1.jpg';
import image2 from './images/2.jpg';
import image3 from './images/3.jpg';
import image4 from './images/4.jpg';
import image5 from './images/5.jpg';

export const images: Record<TImageKeys, string> = {
  '1.jpg': image1,
  '2.jpg': image2,
  '3.jpg': image3,
  '4.jpg': image4,
  '5.jpg': image5,
};
