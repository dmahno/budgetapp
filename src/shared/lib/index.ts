export {formatNumber} from './formatNumber/formatNumber';
export {parseNumber} from './parseNumber/parseNumber';
