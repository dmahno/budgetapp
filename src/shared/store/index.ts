export * from './AuthStore/AuthStore';
export * from './ThemeStore/ThemeStore';
