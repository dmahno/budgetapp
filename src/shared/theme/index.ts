export {configDarkTheme} from './tokens/configDarkTheme';
export {configLightTheme} from './tokens/configLightTheme';
