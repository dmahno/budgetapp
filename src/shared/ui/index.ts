export {Logo} from './Logo/Logo';
export {ReactRouterCustomLink} from './ReactRouterCustomLink/ReactRouterCustomLink';
export {DynamicIcon} from './DynamicIcon/DynamicIcon';
export {TapButton} from './TapButton/TapButton';
