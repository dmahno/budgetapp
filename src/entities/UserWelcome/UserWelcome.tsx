import {observer} from 'mobx-react-lite';

import {authStore} from 'shared/store';

export const UserWelcome = observer(() => {
  if (!authStore.user) {
    return null;
  }

  return (
    <div>
      <h2>Добро пожаловать, {authStore.user.userNameInfo}!</h2>
      {authStore.user.avatar && (
        <img
          src={`../../shared/assets/${authStore.user?.avatar}`}
          alt="User Avatar"
        />
      )}
    </div>
  );
});
