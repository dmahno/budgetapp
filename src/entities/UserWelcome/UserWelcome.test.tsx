import {describe, it, expect, vi, beforeEach} from 'vitest';
import {render, screen} from '@testing-library/react';

import {authStore} from 'shared/store';

import {UserWelcome} from './UserWelcome';

vi.mock('shared/store', () => ({
  authStore: {
    user: null,
  },
}));

describe('UserWelcome', () => {
  beforeEach(() => {
    authStore.user = null;
  });

  it('renders nothing when no user is logged in', () => {
    render(<UserWelcome />);
    expect(screen.queryByText(/Добро пожаловать/i)).not.toBeInTheDocument();
  });

  it('renders welcome message and avatar when user is logged in with avatar', () => {
    authStore.user = {
      id: 1,
      userNameInfo: 'Дмитрий',
      avatar: '1.jpg',
    };

    render(<UserWelcome />);
    expect(screen.getByText('Добро пожаловать, Дмитрий!')).toBeInTheDocument();
    const avatar = screen.getByRole('img');
    expect(avatar).toHaveAttribute('src', '../../shared/assets/1.jpg');
    expect(avatar).toHaveAttribute('alt', 'User Avatar');
  });
  it('matches the snapshot', () => {
    const {asFragment} = render(<UserWelcome />);
    expect(asFragment()).toMatchSnapshot();
  });
});
