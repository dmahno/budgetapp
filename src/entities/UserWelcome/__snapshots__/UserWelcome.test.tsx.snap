// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserWelcome > matches the snapshot 1`] = `<DocumentFragment />`;
